node-websocket-chat/README.md

```diff
@@ -0,0 +1,200 @@
+The complete client code is [here](index.html).
+
+## Extending the Server
+
+Modifications to the server are based on [this article](https://objectcomputing.com/resources/publications/sett/custom-bridges-between-restful-web-services-and-dds), [this repository](https://github.com/OpenDDS/node-opendds-rest-demo), and the [OpenDDS module for Node.js](https://github.com/OpenDDS/node-opendds).
+
+The first step in extending the server is defining topics for the messages the server instances will exchange.
+The name of the topic for this application is `User Messages`.
+The data type used for that topic is named `Chat::UserMessage` and is defined in [idl/Chat.idl](idl/Chat.idl).
+The `Chat::UserMessage` is a struct containing a single string named `message`.
+Both of these names appear in the [chat.js](chat.js) file that will be explained next.
+The IDL file compiles to a shared library that will be loaded by the server.
+
+The second step is wrapping OpenDDS in a way that is convenient for the application.
+The [chat.js](chat.js) file contains a number of helper methods
+
+* `initializeDds` initializes OpenDDS, loads the IDL shared library, creates a DDS Participant, and creates the DataWriter for user messages.
+* `finalizeDds` deletes the participant and finalizes OpenDDS.
+* `writer_user_message` writes a `Chat::UserMessage` to the `User Messages` topic.
+* `subscribe_user_messages` takes a callback that will be invoked whenever a user message is received.  This code enables reliable data transfer which is off by default for a subscriber.
+
+The [server.js](server.js) file is based on the server code for https://karlhadwen.medium.com/node-js-websocket-tutorial-real-time-chat-room-using-multiple-clients-44a8e26a953e.
+The code now contains logic for handling the listening port of the web server so multiple servers can run on the same host and the necessary logic for introducing the OpenDDS wrapper and initializing OpenDDS.
+
+The major change is the code that processes messages received via websocket.
+In the original server code, a message that was received on a websocket connect was sent to all of the connection except the one that originally sent the message.
+This code has been broken into two separate pieces:  one for receiving and one for sending.
+The code that handles messages from the websockets writes the received message to the `User Messages` topic.
+When initializing OpenDDS, the code installs a callback that sends a message received on the `User Messages` topic to all of the websocket clients that are connected.
+Together, these changes result in the following flow for a message:
+
+1. The client sends a message.
+2. The server receives the message via websocket and writes it to the `User Messages` topic.
+3. The server receives the message via the `User Messages` topic and writes it to all connected websocket clients.
+4. The client receives the message from the websocket and displays it.
```

Comment:
I expected this part to describe how the server knows not to send the message back to the client that originated it

Reply:
The server does send the message back to the client.  The system becomes simpler without that optimization.

Reply:
In the part above, how about rewording "The server was optimized to not send a message back to the client who sent it." -- this was unclear to me.  I thought "The server" here is the server you wrote.